localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -819,67 +822,91 @@ std::array<double, 36> NDTScanMatcher::estimate_covariance(
-    const Eigen::Matrix4f sub_ndt_result = ndt_ptr_->getResult().pose;
-
-    const Eigen::Vector2d sub_ndt_pose_2d = sub_ndt_result.topRightCorner<2, 1>().cast<double>();
-    mean += sub_ndt_pose_2d;
-    ndt_pose_2d_vec.emplace_back(sub_ndt_pose_2d);
-
-    multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(sub_ndt_result));
-    multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(sub_initial_pose_matrix));
-  }
-
-  // calculate the covariance matrix
-  mean /= n;
-  Eigen::Matrix2d pca_covariance = Eigen::Matrix2d::Zero();
-  for (const auto & temp_ndt_pose_2d : ndt_pose_2d_vec) {
-    const Eigen::Vector2d diff_2d = temp_ndt_pose_2d - mean;
-    pca_covariance += diff_2d * diff_2d.transpose();
+  if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::LAPLACE_APPROXIMATION) {
+    Eigen::Matrix2d cov_by_la =
+      pclomp::estimate_xy_covariance_by_Laplace_approximation(ndt_result.hessian);
+    const Eigen::Matrix2d cov_by_la_adjust =
+      pclomp::adjust_diagonal_covariance(cov_by_la, ndt_result.pose, 0.0225, 0.0225);
+    ndt_covariance[0 + 6 * 0] = cov_by_la_adjust(0, 0);
+    ndt_covariance[1 + 6 * 1] = cov_by_la_adjust(1, 1);
+    ndt_covariance[1 + 6 * 0] = cov_by_la_adjust(1, 0);
+    ndt_covariance[0 + 6 * 1] = cov_by_la_adjust(0, 1);
+    const Eigen::Matrix2d cov_by_la_rotated =
+      pclomp::rotate_covariance_to_base_link(cov_by_la_adjust, ndt_result.pose);
+    ndt_dev_x_pub_->publish(
+      make_float32_stamped(sensor_ros_time, std::sqrt(cov_by_la_rotated(0, 0))));
+    ndt_dev_y_pub_->publish(
+      make_float32_stamped(sensor_ros_time, std::sqrt(cov_by_la_rotated(1, 1))));
+  } else if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::MULTI_NDT) {
+    const std::vector<Eigen::Matrix4f> poses_to_search = pclomp::propose_poses_to_search(
+      ndt_result, param_.covariance.covariance_estimation.initial_pose_offset_model_x,
+      param_.covariance.covariance_estimation.initial_pose_offset_model_y);
+    const pclomp::ResultOfMultiNdtCovarianceEstimation result_of_multi_ndt_covariance_estimation =
+      estimate_xy_covariance_by_multi_ndt(ndt_result, ndt_ptr_, poses_to_search);
+    const Eigen::Matrix2d cov_by_mndt = pclomp::adjust_diagonal_covariance(
+      result_of_multi_ndt_covariance_estimation.covariance, ndt_result.pose, 0.0225, 0.0225);
+    ndt_covariance[0 + 6 * 0] = cov_by_mndt(0, 0);
+    ndt_covariance[1 + 6 * 1] = cov_by_mndt(1, 1);
+    ndt_covariance[1 + 6 * 0] = cov_by_mndt(1, 0);
+    ndt_covariance[0 + 6 * 1] = cov_by_mndt(0, 1);
+    const Eigen::Matrix2d cov_by_mndt_rotated =
+      pclomp::rotate_covariance_to_base_link(cov_by_mndt, ndt_result.pose);
+    ndt_dev_x_pub_->publish(
+      make_float32_stamped(sensor_ros_time, std::sqrt(cov_by_mndt_rotated(0, 0))));
+    ndt_dev_y_pub_->publish(
+      make_float32_stamped(sensor_ros_time, std::sqrt(cov_by_mndt_rotated(1, 1))));
```

Comment:
Since the processes such as setting the covariance matrix and publishing stddev are common, wouldn't it be better to implement them outside the if statement?

Reply:
I agree with your opinion, thank you for pointing that out.